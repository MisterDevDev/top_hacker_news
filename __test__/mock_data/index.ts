import path from 'path'
import { promises as fs } from "fs";
import { IComment, IStory } from '@/typings';

const JSON_FILEPATH = path.join(__dirname, "json");

export async function readFakeData() {
    const [fakeStories, fakeIds, fakeComment] = await Promise.all([
        getJSONfromFile(filenames.stories),
        getJSONfromFile(filenames.storyIds),
        getJSONfromFile(filenames.comment),
    ])

    return {
        fakeStories: fakeStories as IStory[],
        fakeIds: fakeIds as number[],
        fakeComment:fakeComment as IComment
    }
}

enum filenames {
    stories = "stories.json",
    storyIds = "storyIds.json",
    comment = "comment.json"
  }

  type JsonDataType =
  | IStory[]
  | number[]
  | IComment

    async function getJSONfromFile<ItemType extends JsonDataType>(
    filename: filenames,
  ): Promise<ItemType> {
    const filePath = path.join(JSON_FILEPATH , filename);
    const data = await fs.readFile(filePath);
    return JSON.parse(data.toString());
  }