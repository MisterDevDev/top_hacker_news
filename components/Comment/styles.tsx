import styled from "styled-components";

interface ICommentContainerProps {
  loading: number;
}

export const StyledCommentContainer = styled.div<ICommentContainerProps>`
  padding: 10px;
  font-weight: normal;
  font-size: 14px;
  background-color: rgb(32, 28, 82);
  border-radius: 1rem;
  margin: 5px;
  word-wrap: break-word;

  transition: all 0.3s ease-in;

  max-height: ${props => {
    return props.loading ? `10vh` : "auto";
  }};

  opacity: ${props => {
    return props.loading ? `40%` : "100%";
  }};
`;

export const StyledParsedSpan = styled.span`
  pre {
    white-space: pre-wrap;
    white-space: -moz-pre-wrap;
    white-space: -pre-wrap;
    white-space: -o-pre-wrap;
    word-wrap: break-word;
  }
`;

export const StyledReplyContainer = styled.div`
  margin-left: 10px;
`;

export const StyledHoverSpan = styled.span`
  &:hover {
    cursor: pointer;
    text-decoration: underline;
    text-decoration-color: #e1e1e1;
    text-decoration-thickness: 0.5px;
  }
`;

export const StyledReplyLink = styled.div`
  text-align: right;
  padding-bottom: 1em;
  padding-right: 1em;
`;

export const StyledAuthorContainer = styled.div`
  padding-bottom: 5px;
  font-size: 0.8em;
`;

export const StyledAuthorSpan = styled.span`
  color: #e9967a;
  font-weight: bold;
  font-size: 1.4em;
`;
