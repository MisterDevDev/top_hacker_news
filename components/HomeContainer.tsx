import { story } from "@/typings";

interface Props {
  topStories: story[];
}

export const HomeContainer = (props: Props) => {
  return (
    <main style={{ display: "flex", flexWrap: "wrap" }}>
      {props.topStories.map((story: story) => {
        return (
          <div key={story.id} style={{ margin: "1rem" }}>
            <ul style={{ overflow: "hidden" }}>
              <li>by: {story.by}</li>
              <li>{story.title}</li>
              <li>Comments: {story.descendants}</li>
              <li>Score: {story.score}</li>
              <li>{story.url}</li>
              <li>Time: {story.time}</li>
            </ul>
          </div>
        );
      })}
    </main>
  );
};
