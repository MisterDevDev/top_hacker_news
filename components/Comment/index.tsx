import { IComment } from "@/typings";
import { useState, useEffect } from "react";
import HTMLparse from "html-react-parser";
import * as Styles from "./styles";
import dayjs from "dayjs";
import relativeTime from "dayjs/plugin/relativeTime";
import { CommentsContainer } from "../CommentsContainer";
import { IoIosReturnRight } from "react-icons/io";
import { CgCornerUpRight } from "react-icons/cg";

dayjs.extend(relativeTime);

interface Props {
  kid_id: number;
  isNestedReply: boolean;
}

export const Comment = (props: Props) => {
  const [comment, setComment] = useState<IComment | undefined>();
  const [showReplies, setShowReplies] = useState(false);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    (async () => {
      setLoading(true);
      const request = `https://hacker-news.firebaseio.com/v0/item/${props.kid_id}.json?print=pretty`;
      const response = await fetch(request);
      setComment(await response.json());
      setLoading(false);
    })();
  }, [props.kid_id]);

  return <>{renderComment()}</>;

  function renderComment() {
    if (comment?.deleted) return <>{renderDeletedComment()}</>;
    if (loading) return <>{renderLoading()}</>;
    return (
      <>
        <Styles.StyledCommentContainer loading={loading ? 1 : 0}>
          {renderAuthor()}
          {renderText()}
        </Styles.StyledCommentContainer>
        {renderReplies()}
      </>
    );
  }

  function renderAuthor() {
    return (
      <Styles.StyledAuthorContainer>
        <Styles.StyledAuthorSpan>{comment?.by ?? ""}</Styles.StyledAuthorSpan> {" " + getTime()}
      </Styles.StyledAuthorContainer>
    );
  }

  function renderText() {
    return <Styles.StyledParsedSpan>{HTMLparse(comment?.text ?? "")}</Styles.StyledParsedSpan>;
  }

  function getTime() {
    if (Number(comment?.time ?? 0)) {
      return dayjs.unix(comment?.time ?? 0).fromNow();
    }
    return "";
  }

  function renderReplies() {
    if (comment?.kids?.[0])
      return (
        <>
          {!showReplies && (
            <Styles.StyledReplyLink>
              <IoIosReturnRight size={18} />
              <Styles.StyledHoverSpan onClick={() => setShowReplies(true)}>
                {comment.kids.length + getReply()}
              </Styles.StyledHoverSpan>
            </Styles.StyledReplyLink>
          )}
          {showReplies && !props.isNestedReply && (
            <Styles.StyledReplyLink>
              <CgCornerUpRight size={18} />
              <Styles.StyledHoverSpan onClick={() => setShowReplies(false)}>
                {"Close " + getReply()}
              </Styles.StyledHoverSpan>
            </Styles.StyledReplyLink>
          )}
          {showReplies && (
            <Styles.StyledReplyContainer>
              <CommentsContainer isReply kids={comment.kids} />
            </Styles.StyledReplyContainer>
          )}
        </>
      );
    return <></>;
  }

  function getReply() {
    return (comment?.kids?.length ?? 0) > 1 ? " Replies" : " Reply";
  }

  function renderDeletedComment() {
    <Styles.StyledCommentContainer loading={loading ? 1 : 0}>
      <span>
        <i>Post Deleted</i>
      </span>
    </Styles.StyledCommentContainer>;
  }

  function renderLoading() {
    <Styles.StyledCommentContainer loading={loading ? 1 : 0}>
      <span>Loading...</span>
    </Styles.StyledCommentContainer>;
  }
};
