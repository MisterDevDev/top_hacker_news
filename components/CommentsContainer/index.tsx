import { useState } from "react";
import { Comment } from "../Comment";
import * as Styles from "./styles";
import { RxArrowUp } from "react-icons/rx";

interface Props {
  kids: number[];
  isReply?: boolean;
}

export const CommentsContainer = (props: Props) => {
  const [visibleComments, setVisibleComments] = useState(0);

  return <>{renderCommentsContainer()}</>;

  function renderCommentsContainer() {
    return (
      <Styles.StyledContentContainer>
        {props.kids.map((kid_id: number, idx) => {
          if (idx <= visibleComments)
            return <Comment key={kid_id} kid_id={kid_id} isNestedReply={!!props.isReply} />;
        })}
        {visibleComments < props.kids.length - 1 && renderMoreLink()}
      </Styles.StyledContentContainer>
    );
  }

  function renderMoreLink() {
    if (props.isReply) return renderMoreReplies();
    return renderMoreComments();
  }

  function renderMoreReplies() {
    return (
      <Styles.StyledHoverSpan
        style={{ paddingLeft: "20px" }}
        onClick={() => setVisibleComments(visibleComments + 3)}
      >
        <RxArrowUp size={12} /> More Replies
      </Styles.StyledHoverSpan>
    );
  }

  function renderMoreComments() {
    return (
      <Styles.StyledMoreComments style={{ textAlign: "center" }}>
        <Styles.StyledHoverSpan
          style={{ paddingLeft: "10px" }}
          onClick={() => setVisibleComments(visibleComments + 3)}
        >
          More Comments
        </Styles.StyledHoverSpan>
      </Styles.StyledMoreComments>
    );
  }
};
