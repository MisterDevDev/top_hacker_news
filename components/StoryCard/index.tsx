import { IStory } from "@/typings";
import * as Styles from "./styles";
import dayjs from "dayjs";
import relativeTime from "dayjs/plugin/relativeTime";
import { IoMdEye, IoMdClose } from "react-icons/io";
import { RxEyeClosed } from "react-icons/rx";
import { useState } from "react";
import HTMLparse from "html-react-parser";
import { CommentsContainer } from "../CommentsContainer";

dayjs.extend(relativeTime);

interface Props {
  story: IStory;
}

export const StoryCard = (props: Props) => {
  const [showText, setShowText] = useState(false);
  const [showComments, setShowComments] = useState(false);
  const [eyeIsHovered, setEyeIsHovered] = useState(false);

  return <>{renderStoryCard()}</>;

  function renderStoryCard() {
    return (
      <>
        <Styles.StyledContentContainer>
          {renderHeader()}
          {renderStoryText()}
          {!showText && renderStory()}
        </Styles.StyledContentContainer>
        {!showText && renderComments()}
      </>
    );
  }

  function renderStoryText() {
    return (
      <>
        <Styles.StyledTextContainer isOpen={showText}>
          {HTMLparse(props.story.text ?? "")}
        </Styles.StyledTextContainer>
      </>
    );
  }

  function renderStory() {
    return (
      <>
        {renderBody()}
        {renderFooter()}
      </>
    );
  }

  function renderHeader() {
    const websiteText = getWebsite();
    return (
      <Styles.StyledStoryHeader>
        <Styles.StyledStoryBy>
          {props.story.score + " points by "}
          <Styles.StyledAuthorSpan>{props.story.by}</Styles.StyledAuthorSpan> <br />
          {websiteText && "(" + websiteText + ")"}
        </Styles.StyledStoryBy>
        {props.story.text && renderTextIcon()}
      </Styles.StyledStoryHeader>
    );
  }

  function renderTextIcon() {
    return showText ? (
      <Styles.StyledHoverSpan>
        <IoMdClose size={18} onClick={() => setShowText(false)} />
      </Styles.StyledHoverSpan>
    ) : (
      <Styles.StyledHoverSpan
        onMouseEnter={() => setEyeIsHovered(true)}
        onMouseLeave={() => setEyeIsHovered(false)}
      >
        {eyeIsHovered ? (
          <IoMdEye size={18} onClick={() => setShowText(true)} />
        ) : (
          <RxEyeClosed size={18} onClick={() => setShowText(true)} />
        )}
      </Styles.StyledHoverSpan>
    );
  }

  function renderBody() {
    if (!(props.story.url?.trim() ?? null))
      return <Styles.StyledStoryTitle>{props.story.title}</Styles.StyledStoryTitle>;
    return (
      <Styles.StyledStoryTitle onClick={() => openWebsite()}>
        <Styles.StyledHoverSpan>{props.story.title}</Styles.StyledHoverSpan>
      </Styles.StyledStoryTitle>
    );
  }

  function renderFooter() {
    return (
      <Styles.StyledStoryFooter isOpen={showComments}>
        <Styles.StyledHoverSpan onClick={() => setShowComments(!showComments)}>
          {showComments ? "Close Comments" : "View " + props.story.descendants + " Comments"}
        </Styles.StyledHoverSpan>
        <span aria-label="posted-time">{getTime()}</span>
      </Styles.StyledStoryFooter>
    );
  }

  function renderComments() {
    return showComments && props.story.kids?.[0] && <CommentsContainer kids={props.story.kids} />;
  }

  function getTime() {
    const relativeStoryTime = dayjs.unix(props.story.time).fromNow();
    return relativeStoryTime.includes("days")
      ? dayjs.unix(props.story.time).format("MMM DD, YYYY")
      : relativeStoryTime;
  }

  function getWebsite() {
    try {
      const urlString = new URL(props.story?.url ?? "").hostname;
      return urlString.replace(/^www\./, "").trim();
    } catch (error) {
      return "";
    }
  }

  function openWebsite() {
    if (props.story.url) {
      window.open(props.story.url);
    }
  }
};
