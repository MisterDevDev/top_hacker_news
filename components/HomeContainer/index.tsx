import { IStory } from "@/typings";
import * as Styles from "./styles";
import { StoryCard } from "../StoryCard";
import { HeroParticles } from "./Particles";
import { useState, useRef, useEffect } from "react";
import useObserveScreen from "./useObserveScreen";

interface Props {
  topStories: IStory[];
}

export const HomeContainer = (props: Props) => {
  const [visibleStories, setVisibleStories] = useState(10);
  const lastStory = useRef(null);

  const isVisible = useObserveScreen(
    {
      root: null,
      rootMargin: "0px",
      threshold: 0.3,
    },
    lastStory
  );

  useEffect(() => {
    if (isVisible) addVisibleStories();
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [isVisible]);

  return (
    <main>
      {renderTitleSection()}
      {renderStoryContainer()}
    </main>
  );

  function addVisibleStories() {
    if (visibleStories <= 90) setVisibleStories(visibleStories + 10);
  }

  function renderTitleSection() {
    return (
      <>
        <HeroParticles />
        <Styles.StyledHeroArea>
          <Styles.StyledSubtitleBox>
            <Styles.StyledTitle>Hacker News</Styles.StyledTitle>
            <Styles.StyledGradientText>
              <i>Top 100</i>
            </Styles.StyledGradientText>
          </Styles.StyledSubtitleBox>
        </Styles.StyledHeroArea>
      </>
    );
  }

  function renderStoryContainer() {
    return <Styles.StyledStoryContainer>{renderStories()}</Styles.StyledStoryContainer>;
  }

  function renderStories() {
    return (
      <>
        {props.topStories.map((story: IStory, idx) => {
          if (story.deleted) return <></>;
          if (idx + 1 === visibleStories) {
            return (
              <Styles.StyledStoryCard key={story.id}>
                <div ref={lastStory} aria-label="news-story">
                  <StoryCard story={story} />
                </div>
              </Styles.StyledStoryCard>
            );
          }
          if (idx + 1 < visibleStories) {
            return (
              <Styles.StyledStoryCard key={story.id}>
                <div aria-label="news-story">
                  <StoryCard story={story} />
                </div>
              </Styles.StyledStoryCard>
            );
          }
        })}
      </>
    );
  }
};
