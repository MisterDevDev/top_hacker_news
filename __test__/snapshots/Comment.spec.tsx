import React from "react";
import { waitFor } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Comment } from "@/components/Comment";
import { readFakeData } from "../mock_data";
import fetchMock from "jest-fetch-mock";
import renderer from "react-test-renderer";

it("Should display the number of replies", async () => {
  const { fakeComment } = await readFakeData();
  fetchMock.mockOnce(JSON.stringify(fakeComment));

  const tree = renderer.create(<Comment kid_id={fakeComment.id} />);

  await waitFor(() => {
    expect(tree.toJSON()).toMatchSnapshot();
    tree.unmount();
  });
});
