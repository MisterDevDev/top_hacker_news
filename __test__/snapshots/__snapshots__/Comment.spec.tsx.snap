// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should display the number of replies 1`] = `null`;
